contrib/cmake/build3rdParty/build3rdParty.cpp

```diff
@@ -0,0 +1,273 @@
+                }
+                if (portversion.size() == 40)
+                {
+                    fs::current_path(cloneDir);
+                    execute("git checkout --quiet " + portversion);
+                    cout << "Copying port for " << portname << " from vcpkg commit " << portversion << endl;
+                    fs::copy(cloneDir / "ports" / portname, vcpkgDir / "ports" / portname, fs::copy_options::recursive);
+                    fs::current_path(vcpkgDir);
+                }
+                else
+                {
+                    cout << "Copying port for " << portname << " from SDK customized port " << portversion << endl;
+                    fs::copy(sdkRootPath / "contrib" / "cmake" / "vcpkg_extra_ports" / portname / portversion, vcpkgDir / "ports" / portname, fs::copy_options::recursive);
+                }
+            }
+
+            if (removeUnusedPorts)
+            {
+                for (auto dir = fs::directory_iterator(vcpkgDir / "ports"); dir != fs::directory_iterator(); ++dir)
+                {
+                    if (ports.find(dir->path().filename().u8string()) == ports.end())
+                    {
+                        fs::remove_all(dir->path());
+                    }
+                }
+            }
+
+        }
+        else if (build)
+        {
+
+            if (!fs::is_directory("vcpkg"))
+            {
+                cout << "This command should be run from just outside 'vcpkg' folder - maybe it is not set up?" << endl;
+                return 1;
+            }
+            else
+            {
+                fs::current_path("vcpkg");
+            }
+
+
+            for (auto portPair : ports)
+            {
+                #ifdef WIN32
+                    execute("vcpkg install --triplet " + triplet + " " + portPair.first);
+                #else
+                    execute("./vcpkg install --triplet " + triplet + " " + portPair.first);
+                #endif
+            }
+        }
+
+
+    }
+
+    return 0;
+}
+catch (exception& e)
+{
+    cout << "exception: " << e.what() << endl;
+    return 1;
+}
+
+
+void execute(string command)
+{
+    cout << "Executing: " << command << endl;
+    int result = system(command.c_str());
+    if (result != 0)
+    {
+        cout << "Command failed with resuld code " << result << " ( command was " << command << ")" <<endl;
+        exit(1);
+    }
+}
+
+
+bool showSyntax()
+{
+    cout << "build3rdParty --setup [--removeunusedports] --ports <ports override file> --triplet <triplet> --sdkroot <path>" << endl;
+    cout << "build3rdParty --build --ports <ports override file> --triplet <triplet>" << endl;
+    return false;
+}
+
+bool readCommandLine(int argc, char* argv[])
+{
+    if (argc <= 1) return showSyntax();
+
+    std::vector<char*> myargv1(argv + 1, argv + argc);
+    std::vector<char*> myargv2;
+
+    for (auto it = myargv1.begin(); it != myargv1.end(); ++it)
+    {
+        if (std::string(*it) == "--ports")
```

Comment:
We might want to do `const std::string currentArg(*it)` or similar; in the case we hit the final `else ` at the bottom, we've constructed the same temporary string six times. 

Also untested, maybe overkill, but maybe more maintainable:

```cpp
using namespace std::string_literals;

map<string, reference_wrapper<bool> flags{
  {"--setup"s, ref(setup)},
  {"--removeUnusedPorts"s, ref(removeUnusedPorts)},
  {"--build"s, ref(build)}
};

map<string, reference_wrapper<string>> options{
 {"--ports"s, ref(portsFile)},
 // ...
};

for (auto it = myargv1.begin(); it != myargv1.end(); ++it) {
  const string currentArg(*it);
  auto flagIt = flags.end();
  auto optIt = options.end();

  if ((flagIt = flags.find(currentArg)) != flags.end()) {
    if (++it == myargv1.end()) return showSyntax();
    flagIt->second = *it;
  } else if ((optIt = options.find(currentArg)) != options.end()) {
    // same check
    optIt->second = true;
  } else {
    myargv2.push_back(*it);
  }
}
```

Reply:
For this small program where performance is unimportant, I'd lean towards keeping the code small rather than trying to be efficient.